Add other actions on VMs from the dashboard

ageorgou opened the pull request:
This adds functionality to #43 and ~will resolve~ addresses most of #9. I don't know whether it's easier to compare this to #43 or to the development branch.

Still work in progress. The basic outline is:
- [x] Get the status of a machine hosted on Azure by querying Azure rather than the DB.
- [x] Add tasks for starting, stopping and restarting a machine.
- [x] Refresh the DB status automatically, or think of some other way to ensure it is up-to-date (or at least base decisions on the live status, not the stored one).
  - Note that this is particularly strange when starting to deploy a new host, as the status is showing as "Failed to deploy" (since the VM does not exist yet), but the deployment is continuing normally!

Deferred for future PR:
- ~[ ] Implement checks for when actions are applicable/allowed.~
- ~[ ] Always show all buttons on the dashboard, with greyed-out versions for inapplicable actions.~
- ~[ ] Other actions mentioned in #9.~
- ~[ ] Tests for new actions.~

ageorgou committed:
Add placeholder for getting status

ageorgou committed:
Add initial plan for getting status

This is quite hacky and I don't like it, but it should work. Getting
the status on Azure is not straightforward...

ageorgou committed:
Finish a rough version of getting status

ageorgou committed:
Display 'live' status instead of checking the DB

ageorgou committed:
Update requirements with Azure SDK

ageorgou committed:
Improve PEP8 compliance

ageorgou committed:
Remove debug prints, oops

ageorgou committed:
Move Azure-related functionality to its own module

ageorgou committed:
Add celery task for stopping a VM

ageorgou committed:
Clarify that Azure 'stopped' status is not yet off

A stopped (but not deallocated) VM still incurs charges, whereas a
dellocated one does not. It would therefore be misleading to show
a stopped machine as actually stopped, especially since the status
could then move back to 'stopping' (while deallocating). I am not
sure whether deallocating happens as a second step or is a wholly
separate operation, but in either case I think it should be fine
to treat Azure 'stopped' as a temporary status, i.e. map it to our
'stopping' Host status.

ageorgou committed:
Launch stop task from dashboard

ageorgou committed:
Add task for starting a stopped VM

ageorgou commented:
Something I forgot to put in the commit message but is worth noting: the version requirement on the Azure SDK packages is because some previous versions do not have all the API methods/fields (in particular `instance_views` was missing, I think). This is the version I've been trying and works.

ageorgou committed:
Add status for DB entries not found in the cloud

In case e.g. a host has been destroyed, but hasn't been deleted
from the database yet. Currently only used when displaying the
status in the dashboard. Fixes #57.

ageorgou committed:
Update the DB with each VM's live status

Added a periodic task that queries Azure and updates the existing
hosts' statuses in the DB. We will need to run `celery beat` for
this, in addition to starting the Celery workers. At the moment,
the dashboard displays the live status but the buttons depend on
the DB status (which will be updated eventually, anyway). Ideally,
we want some logic to combine the live status with info from the
application, particularly to identify hosts being deployed (which
we cannot do with information from Azure only).

ageorgou committed:
Ignore live status for hosts still in deployment

ageorgou committed:
Show database status on dashboard

Now that the databse is updated in the background, we can show that
status on the dashboard instead of the live one.

ageorgou committed:
Store Terraform template in database

ageorgou committed:
Create list of current deployments

ageorgou committed:
Delete deploying hosts through Azure SDK

ageorgou committed:
Rerender template if not found in DB when deleting

raquelalegre committed:
Update to latest version of Azure's TF plugin.

raquelalegre committed:
Azure changed its way to define a vnet and it is now mandatory to include a reference from the vnet definition to the subnet.

ageorgou committed:
Remove empty line

ageorgou committed:
Merge branch 'bugfix/terraform_vnet' into destroy-manage

ageorgou committed:
Store task ID of new deployments in DB

raquelalegre commented on requirements/base.in:
```diff
@@ -17,3 +17,7 @@ celery
 haikunator
+
+# Azure
+azure-mgmt-compute>=3.0.1
```
Comment on why versions < 3.0.1 are not suitable?

ageorgou replied:
Sorry, mentioned it above but I thought it was in the code too: earlier package versions do not provide the full API. It's possible that some older versions are also OK, but I went with the one I was trying for sure.

raquelalegre commented on src/cloudlabs/app.py:
```diff
@@ -21,6 +21,7 @@ def create_app(config_name=None):
         import os
         config_name = os.getenv('APP_SETTINGS', 'cloudlabs.config.Config')
     app = Flask(__name__)
+    app.current_deployments = {}
```
Not needed anymore.

raquelalegre commented on src/cloudlabs/azure_tools.py:
```diff
@@ -0,0 +1,44 @@
+
+    Also see: https://github.com/Azure-Samples/virtual-machines-python-manage
+    """
+    # TODO are secrets better accessible in a different way?
```
If I remember well, the Azure secrets were in environment vars and then were moved to the `Secrets` object, so I guess this TODO item was here before the changes starting in line 22. So delete lines 15-21.

raquelalegre commented on src/cloudlabs/azure_tools.py:
```diff
@@ -0,0 +1,44 @@
+    return credentials, subscription_id
+
+
+def get_compute_manager(credentials=None, subscription_id=None):
```
Wouldn't it be nicer to make this utils into a class and then refer to credentials and subscription id as member vars instead of passing them around? That way the code that uses this doesn't need to retrieve and pass around arguments.

raquelalegre commented on src/cloudlabs/context_processors.py:
```diff
@@ -20,3 +19,9 @@ def inject_status():
         return {'HostStatus': HostStatus}
+
+    @app.context_processor
+    def inject_get_status():
```
You mentioned this was not needed, so just recording that here because I can't tell! :)

raquelalegre commented on src/cloudlabs/deployer/deployer.py:
```diff
@@ -151,3 +161,44 @@ def destroy(self, host):
+        action.wait()
+        # TODO Record result?
+
+    def restart(self, host):
```
As per office discussion we are not sure if Azure's `start` does a restart or it's needed that we have an extra start method here.

ageorgou committed:
Add the stopping host status (somehow missed!)

ageorgou committed:
Comment on new celery details

raquelalegre commented on src/cloudlabs/tasks/worker.py:
```diff
@@ -29,21 +38,27 @@ def deploy(host_id):
 @celery.task(name='cloudlabs.destroy')
-def destroy(host_id):
+def destroy(host_id, hard):
+    print("Destroying {} from task {}".format(host_id, deploy.request.id))
```
This is probably a leftover debug message that should be a log message when we implement a logger for CloudLabs in a separate PR.

raquelalegre reviewed (changes requested):
It all seems fine and works on my machine, so just minor changes requested. It'd be nice, but better on a separate PR, to have some more log messages. That'll also give us some information about how the system is being used so we are able to trace problems. I'll create a new issue.

ageorgou committed:
Clean up redundant code

As per PR comments. I've left in the debug messages as comments for
when we implement logging. Also removing another unused import.

ageorgou committed:
Capture Azure functionality in a class

ageorgou committed:
Use new Azure tools class

This makes some of the deployer methods (except for deploy and
destroy) trivial. They could probably be removed and we could
then just have the start, stop etc methods called form the celery
tasks. I am leaving them in for now because we may still want to
do something for which we need the deployer (eg update the DB).
This commit also makes the start/restart methods more consistent
(restart is only for hosts that are already running, start is for
either stopped or running).

ageorgou committed:
Move status retrieval to new tools class

Plus small cleanup and fixes.

ageorgou committed:
Move "deployment unsetting" to right place!

Not sure how that was missed -- we are meant to unset the task
field after a (successful or failed) deployment attempt, not after
a reset!

ageorgou committed:
Enable starting a stopped machine functionality

Previously only restarting and stopping were available.

ageorgou commented:
As per offline discussion, this is enough for this PR. I'll leave remaining tasks (icons, checks for applicable actions, tests?) for another.
